temba/api/v2/views.py

```diff
@@ -5008,40 +5008,45 @@ class MessagesReportEndpoint(BaseAPIView, ReportEndpointMixin):
     def filter_and_paginate_qs_by_flow(self, org, flow, qs):
         # filter and paginate flow runs
         runs = (
-            flow.runs.filter(
-                self.get_datetime_filters("", "exited_on", org),
-                org=org,
-                status__in=[
-                    FlowRun.STATUS_COMPLETED,
-                    FlowRun.STATUS_INTERRUPTED,
-                    FlowRun.STATUS_FAILED,
-                    FlowRun.STATUS_EXPIRED,
-                ],
-            )
-            .only("flow_id", "events", "modified_on")
+            FlowRun.objects.filter(self.get_datetime_filters("", "exited_on", org), flow_id=flow.id)
+            .exclude(status__in=[FlowRun.STATUS_ACTIVE, FlowRun.STATUS_WAITING])
+            .only("events", "modified_on")
             .using("read_only_db")
         )
-        self.configure_paginator_for_flow_runs(runs)
+        self.configure_paginator_for_flow_runs(flow)
         runs, next_page = self.get_paginated_queryset(runs)
 
         # get messages uuids from filtered runs
-        messages_uuids = [
+        messages_uuids = set(
             evt["msg"].get("uuid")
             for run in runs
             for evt in run.get_msg_events()
             if evt["msg"].get("uuid") not in [None, "None", ""]
-        ]
+        )
 
         # filter messages by uuids
         qs = qs.filter(uuid__in=messages_uuids) if messages_uuids else Msg.objects.none()
-        qs = qs.only("created_on").annotate(ds=Concat("direction", "status")).using("read_only_db")
+        qs = qs.only("created_on", "direction", "status").using("read_only_db")
```

Comment:
If we have the `[:len(messages_uuids)]` already on this line, would avoid querying everything, right? Or does not matter?

Reply:
It should stop processing the messages query when all uuids were found.